---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# fixbatch

<!-- badges: start -->
<!-- badges: end -->

## Update

### fix bug 2020-12-24

- [x] add the function to select housekeep gene Separately from every batch group

- [x] add the function to select some housekeep gene that you know its' function or famous stable gene like actinB.

- [x] add the `edger_fix` function to calculate the adjuts expression cpm value form `egdeR`

### fix bug 2020-12-29

- [x] add the `pvca` function form pvca packgas in Bioconductor,this function can help us  to estimate factors' partition of the total variability. I have made some optimizations,such as suppressMessages on useless messages, add test of input data have NAs,fix the the R packages dependency relationship. and add `pvca_plot` plot function to display the result.

- [x] add the `combat_fix` copy form `sva`. IT is exactly the same a sva::combat, only changing name. the sva website is "https://bioconductor.org/packages/release/bioc/html/sva.html"



the R packages is full fo bugs, beacuse it's a naive packagse, so I suggest every time when you use the R packages, should update R packages fristly. Remember this step everytimes!

```{r eval=FALSE}
devtools::update_packages("fixbatch")
```

## Insatll

```{r eval=FALSE}
devtools::install_github("wangjiaxuan/fixbatch")
```

```{r setup}
library(fixbatch)
```

## Function

### col_cv 

calculate the coefficient of variation of the row value in a gene expression matrix.

### col_sum

calculate the sum of rows value in a gene expression matrix.

### housekeep

According to the coefficient of variation and the expression sum, the stable observation value is selected

### batch_housekeep

The function is made for select housekeep gene separately from every batch group

### norm_fix

norm_fix allows users to adjust for batch effects in datasets where the batch covariate is known, using the packages `preprocessCore`, and the function `normalize.quantiles`.

### edger_fix

Calculate normalization factors to scale the raw library sizes. the script is created by Yunshun Chen. The work use the function calcNormFactors and cpm.

### pvca

The function is written based on the 'pvcaBatchAssess' function of the PVCA R package(http://watson.nci.nih.gov/bioc_mirror/packages/release/bioc/manuals/pvca/man/pvca.pdf).
and adjust by Donghyung Lee for changed slightly to make it more efficient and flexible for sequencing read counts data(https://github.com/dleelab/pvca). But it didn't update since 4 years ago.
My new R packages fixbatch need the function to estimate factor's partition of the total variability. So I fork the packages and Adjust in my own style.

### pvca_plot

Title A plot function for dispaly the result of principal variance component analysis

## 开发过程

1. [处理批次效应连续剧第一集（失败的R包）](https://mp.weixin.qq.com/s/_LNdR7b4LRhiKGqIcEXX2A)
2. [处理批次效应连续剧第二集（R包小有成效）](https://mp.weixin.qq.com/s/pWO9em16WE90T-6vxO3HMg)
3. [处理批次效应连续剧第三集（R包基本OK）](https://abego.cn/2020/12/24/remove-the-batch-effect-series-3-r-package-more-power/)


